tsdb/engine/tsm1/engine_test.go

```diff
@@ -1819,28 +1750,168 @@ func BenchmarkEngine_WritePoints_Parallel(b *testing.B) {
+	},
+	{
+		name: "GroupByTime_1m-1h",
+		modify: func(opt query.IteratorOptions) query.IteratorOptions {
+			opt.StartTime = 0
+			opt.EndTime = int64(time.Hour) - 1
+			opt.Interval = query.Interval{
+				Duration: time.Minute,
+			}
+			return opt
+		},
+	},
+	{
+		name: "GroupByTime_1h-1d",
+		modify: func(opt query.IteratorOptions) query.IteratorOptions {
+			opt.StartTime = 0
+			opt.EndTime = int64(24*time.Hour) - 1
+			opt.Interval = query.Interval{
+				Duration: time.Hour,
+			}
+			return opt
+		},
+	},
+	{
+		name: "GroupByTime_1m-1d",
+		modify: func(opt query.IteratorOptions) query.IteratorOptions {
+			opt.StartTime = 0
+			opt.EndTime = int64(24*time.Hour) - 1
+			opt.Interval = query.Interval{
+				Duration: time.Minute,
+			}
+			return opt
+		},
+	},
+	{
+		name: "GroupByHost",
+		modify: func(opt query.IteratorOptions) query.IteratorOptions {
+			opt.Dimensions = []string{"host"}
+			return opt
+		},
+	},
+	{
+		name: "GroupByHostAndTime_1m-1h",
+		modify: func(opt query.IteratorOptions) query.IteratorOptions {
+			opt.Dimensions = []string{"host"}
+			opt.StartTime = 0
+			opt.EndTime = int64(time.Hour) - 1
+			opt.Interval = query.Interval{
+				Duration: time.Minute,
+			}
+			return opt
+		},
+	},
+	{
+		name: "GroupByHostAndTime_1h-1d",
+		modify: func(opt query.IteratorOptions) query.IteratorOptions {
+			opt.Dimensions = []string{"host"}
+			opt.StartTime = 0
+			opt.EndTime = int64(24*time.Hour) - 1
+			opt.Interval = query.Interval{
+				Duration: time.Hour,
+			}
+			return opt
+		},
+	},
+	{
+		name: "GroupByHostAndTime_1m-1d",
+		modify: func(opt query.IteratorOptions) query.IteratorOptions {
+			opt.Dimensions = []string{"host"}
+			opt.StartTime = 0
+			opt.EndTime = int64(24*time.Hour) - 1
+			opt.Interval = query.Interval{
+				Duration: time.Hour,
+			}
+			return opt
+		},
+	},
 }
 
-func benchmarkIterator(b *testing.B, opt query.IteratorOptions, pointN int) {
+func BenchmarkEngine_CreateIterator_Default_1K(b *testing.B) {
+	benchmarkIteratorDefault(b, 1000)
+}
+func BenchmarkEngine_CreateIterator_Default_100K(b *testing.B) {
+	benchmarkIteratorDefault(b, 100000)
+}
+func BenchmarkEngine_CreateIterator_Default_1M(b *testing.B) {
+	benchmarkIteratorDefault(b, 1000000)
+}
+
+func benchmarkIteratorDefault(b *testing.B, pointN int) {
 	e := MustInitDefaultBenchmarkEngine(pointN)
-	b.ResetTimer()
-	b.ReportAllocs()
+	benchmarkEngine(b, e)
+}
 
-	for i := 0; i < b.N; i++ {
-		itr, err := e.CreateIterator(context.Background(), "cpu", opt)
-		if err != nil {
-			b.Fatal(err)
+func benchmarkEngine(b *testing.B, e *Engine) {
+	for _, tt := range benchmarks {
+		for _, variant := range benchmarkVariants {
+			name := tt.name + "_" + variant.name
+			opt := variant.modify(tt.opt)
+			b.Run(name, func(b *testing.B) {
+				b.ReportAllocs()
+				for i := 0; i < b.N; i++ {
+					itr, err := e.CreateIterator(context.Background(), "cpu", opt)
+					if err != nil {
+						b.Fatal(err)
+					}
+					query.DrainIterator(itr)
```

Comment:
Does `DrainIterator` have a significant cost or any allocations? You could surround it in a `stop` / `reset` or whatever if it does.

Reply:
It reads the iterator so all of the logic we're testing happens because that's initiating the calls to `Next()`. So that needs to be part of the profile.